firefox-ios/firefox-ios-tests/Tests/ClientTests/TemporaryDocumentTests.swift

```diff
@@ -0,0 +1,180 @@
+    }
+
+    func testDownload() {
+        let expectation = XCTestExpectation(description: "download callback should be fired")
+        MockURLProtocol.data = Data()
+        subject = createSubject(filename: filename, request: request, session: mockURLSession)
+
+        subject.download { url in
+            XCTAssertNotNil(url)
+            XCTAssertEqual(url?.lastPathComponent, self.filename)
+            XCTAssertTrue(Thread.isMainThread)
+            expectation.fulfill()
+        }
+
+        wait(for: [expectation])
+    }
+
+    func testDownload_onDownloadStarted() {
+        let expectation = XCTestExpectation(description: "onStart callback should be fired")
+        MockURLProtocol.data = Data()
+        subject = createSubject(filename: filename, request: request, session: mockURLSession)
+
+        subject.onDownloadStarted = {
+            expectation.fulfill()
+        }
+        subject.urlSession(
+            mockURLSession,
+            downloadTask: mockURLSession.downloadTask(with: request),
+            didWriteData: 0,
+            totalBytesWritten: 0,
+            totalBytesExpectedToWrite: 0
+        )
+
+        wait(for: [expectation])
+        // it is nilled out so to not call it again
+        XCTAssertNil(self.subject.onDownloadStarted)
+    }
+
+    func testDownload_onProgressUpdate() {
+        let expectation = XCTestExpectation(description: "onDownloadProgressUpdate callback should be fired")
+        let mockProgress = 100.0
+        MockURLProtocol.data = Data()
+        subject = createSubject(filename: filename, request: request, session: mockURLSession)
+
+        subject.onDownloadProgressUpdate = { progress in
+            XCTAssertEqual(progress, mockProgress)
+            expectation.fulfill()
+        }
+        subject.urlSession(
+            mockURLSession,
+            downloadTask: mockURLSession.downloadTask(with: request),
+            didWriteData: 0,
+            totalBytesWritten: Int64(mockProgress),
+            totalBytesExpectedToWrite: 1
+        )
+
+        wait(for: [expectation])
+    }
+
+    func testDeinit_removeTempFile() {
+        XCTAssert(true)
+    }
+
+    // MARK: - URLSessionDownloadDelegate
+
+    func urlSession(_ session: URLSession, downloadTask: URLSessionDownloadTask, didFinishDownloadingTo location: URL) {
+        guard let subject else { return }
+        subject.urlSession(session, downloadTask: downloadTask, didFinishDownloadingTo: location)
+    }
+
+    func urlSession(_ session: URLSession, task: URLSessionTask, didCompleteWithError error: (any Error)?) {
+        guard let subject else { return }
+        subject.urlSession(session, task: task, didCompleteWithError: error)
+    }
+
+    private func createSubject(
+        filename: String?,
+        request: URLRequest,
+        session: URLSession,
+        mimeType: String? = nil,
+        cookies: [HTTPCookie] = []
+    ) -> DefaultTemporaryDocument {
+        return DefaultTemporaryDocument(
```

Comment:
We could use `trackForMemoryLeaks` for `DefaultTemporaryDocument` since it's a class